File: 98-validate-binary-search-tree/98.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Info:
    def __init__(self, isValid=True, minVal=math.inf, maxVal=-math.inf):
        self.isValid = isValid
        self.minVal = minVal
        self.maxVal = maxVal

class Solution:
    def dfs(self, root):
        if root is None:
            return Info()

        left = self.dfs(root.left)
        right = self.dfs(root.right)

        info = Info()
        info.isValid = left.isValid and right.isValid and left.maxVal < root.val and right.minVal > root.val
        info.minVal = min(root.val, left.minVal)
        info.maxVal = max(root.val, right.maxVal)
        return info

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        ret = self.dfs(root)

        return ret.isValid

File: 98-validate-binary-search-tree/98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct TreeInfo {
    long minValue;
    long maxValue;
    bool isValid;
};
class Solution {
private:
    TreeInfo dfs(TreeNode* node) {
        if (!node) {
            return {
                .minValue = 2147483648,
                .maxValue = -2147483649,
                .isValid = true
            };
        }
        TreeInfo left = dfs(node->left);
        TreeInfo right = dfs(node->right);
        long value = node->val;
        return {
            .minValue = min(value, min(left.minValue, right.minValue)),
            .maxValue = max(value, max(left.maxValue, right.maxValue)),
            .isValid = left.isValid && right.isValid && left.maxValue < node->val && right.minValue > node->val
        };
    }
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root).isValid;
    }
};
